src/features/mcda/atoms/mcdaCalculation.ts

```diff
@@ -0,0 +1,223 @@
+  featureProp,
+  notEqual,
+} from '~utils/bivariate/bivariate_style/styleGen';
+import { adaptTileUrl } from '~utils/bivariate/tile/adaptTileUrl';
+import appConfig from '~core/app_config';
+import {
+  MCDARenderer,
+  sentimentDefault,
+  sentimentReversed,
+} from '../renderers/MCDARenderer';
+import type { BivariateLayerStyle } from '~utils/bivariate/bivariateColorThemeUtils';
+import type { Quotient } from '~utils/bivariate';
+
+export interface JsonMCDA {
+  id?: string;
+  axes: [string, string][];
+  ranges: [number, number][];
+  sentiments: [string, string][];
+  coefficients: number[];
+  colors: {
+    good: string;
+    bad: string;
+  };
+}
+
+const DEFAULT_GREEN = 'rgba(90, 200, 127, 0.5)';
+const DEFAULT_RED = 'rgba(228, 26, 28, 0.5)';
+
+const filterSetup = (quotients: Quotient[]) =>
+  anyCondition(
+    ...quotients.map((quotient) =>
+      notEqual(['/', featureProp(quotient[0]), featureProp(quotient[1])], 0),
+    ),
+  );
+
+const layerNormalized = (
+  [num, den]: Quotient,
+  [min, max]: [number, number],
+  sentiment: [string, string],
+  coefficient: number,
+) => {
+  const layerNormalizedFormula = [
+    '*',
+    coefficient,
+    ['/', ['-', ['/', ['get', num], ['get', den]], min], max - min],
+  ];
+
+  if (isEqual(sentiment, sentimentDefault)) {
+    return layerNormalizedFormula;
+  } else if (isEqual(sentiment, sentimentReversed)) {
+    return ['-', 1, layerNormalizedFormula];
+  }
+};
+
+const linearNormalization = (json: JsonMCDA) => {
+  const layersCount = json.axes.length;
+  if (layersCount === 1) {
+    return layerNormalized(
+      json.axes[0],
+      json.ranges[0],
+      json.sentiments[0],
+      json.coefficients[0],
+    );
+  } else if (layersCount > 1) {
+    return [
+      '/',
+      [
+        '+',
+        ...json.axes.map((_, i) =>
+          layerNormalized(
+            json.axes[i],
+            json.ranges[i],
+            json.sentiments[i],
+            json.coefficients[i],
+          ),
+        ),
+      ],
+      sum(json.coefficients),
+    ];
+  }
+};
+
+export const mcdaCalculationAtom = createAtom(
+  {
+    calcMCDA: (json: JsonMCDA) => json,
+    enableMCDALayer: (layerId: string) => layerId,
+    disableMCDALayer: () => null,
+  },
+  ({ onAction, schedule, getUnlistedState, create }) => {
+    onAction('calcMCDA', (json) => {
+      const id = json.id || 'MCDA_layer';
+      const { good = DEFAULT_GREEN, bad = DEFAULT_RED } = json.colors;
+      const layerStyle: BivariateLayerStyle = {
+        id,
+        type: 'fill',
+        layout: {},
+        filter: filterSetup(json.axes),
+        paint: {
+          'fill-color': [
+            'let',
+            'mcdaResult',
+            ['to-number', linearNormalization(json), -1], // falsy values become -1
+            [
+              'case',
+              ['all', ['>=', ['var', 'mcdaResult'], 0], ['<=', ['var', 'mcdaResult'], 1]],
+              ['interpolate', ['linear'], ['var', 'mcdaResult'], 0, bad, 1, good],
```

Comment:
how about interpolate-hcl instead of interpolate? https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#interpolate-hcl